src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/AccessLevel.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2016, 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.javadoc.internal.tool;
+
+import java.util.Set;
+import javax.lang.model.element.Modifier;
+
+/**
+ * The access levels.
+ *
+ * These constants are ordered by their access limiting power. The bigger the
+ * {@link #ordinal() ordinal} of a constant, the more limiting power that
+ * constant has.
+ *
+ * That has a few useful implications. For example, the levels can be compared
+ * by {@link #compareTo}. It also means that {@code AccessLevel.values()[0]} and
+ * {@code AccessLevel.values()[values.length() - 1] and the constants with the
+ * smallest and the biggest limiting powers respectively.
+ */
+public enum AccessLevel {
+
+    /** Does not limit access */
+    PRIVATE,
+    /** Limits access to entities that are public, protected, or declared with package access */
+    PACKAGE,
+    /** Limits access to public and protected entities */
+    PROTECTED,
+    /** Limits access to public entities */
+    PUBLIC;
+
+    static { // self-test to catch unintended reordering of the constants
+        assert PRIVATE.ordinal() == 0
+                && PACKAGE.ordinal() == 1
+                && PROTECTED.ordinal() == 2
+                && PUBLIC.ordinal() == 3;
+    }
```

Comment:
Yet another alternate suggestion would be to use something like

```
assert List.of(AccessLevel.values())
    .equals(List.of(PRIVATE, PACKAGE, PROTECTED, PUBLIC)
```
    
 although I still think it is paranoid  (and non-standard) to assert the order of enum members for any enum that is used as `Comparable`

Reply:
> How much is the order actually relied on?

Clients of `doclint.Env.AccessKind` use `compareTo()`, which is effectively defined through `ordinal()`.

> A different solution, if one is really needed, would be a regression test to verify the expected order.

There are such tests already. For example, DocLintReferencesTest.

> Yet another alternate suggestion would be to use something like
> 
> ```
> assert List.of(AccessLevel.values())
>     .equals(List.of(PRIVATE, PACKAGE, PROTECTED, PUBLIC)
> ```
> 

This is certainly shorter. However, my IDE no longer flags that assert statement as "always true".

> I still think it is paranoid (and non-standard) to assert the order of enum members for any enum that is used as `Comparable`

For better or worse, every enum exposes its constants' declaration order. That order may unexpectedly become relied upon elsewhere, and the author of the enum cannot do anything about it. To me, this is still a gotcha moment.

When I found that doclint.Env.AccessKind.compareTo was used, my initial reaction (after Yuck!) was to introduce explicit construction AccessKind(int level) and boolean instance methods, such as lessLimiting(AccessLevel), equallyLimiting(AccessLevel) and moreLimiting(AccessLevel), that would compare int levels that the constants were constructed with.

But on second thought, it felt like working against the language, which gives us all these features for free. So I decided to embrace that part of enums design and aid the next person who will look at this code, by adding a test, a comment or an assertion.

There were already a few tests, so I decided not to add more. An assertion and a comment together are better than just an assertion, which is better than just a comment. Assertion is a checked comment: it jumps out at you, it fails fast, and works nicely with a sufficiently smart IDE.

Reply:
> For better or worse, every enum exposes its constants' declaration order. That order may unexpectedly become relied upon elsewhere, and the author of the enum cannot do anything about it. T

Since any `enum` implements `Comparable`, it is not reasonable to say that the order may be relied on "unexpectedly".  It is an intentionally defined feature of the language design.

Reply:
While it is not _necessary_ to indicate that an enum implements `Comparable`, it is permissible to state it explicitly, such as in `enum Foo implements Comparable<Foo> { foo1, foo2 }`.  which is a somewhat more linguistic way of writing `/** The order of the constants is important. */ enum Foo { foo1, foo2 }`

Reply:
> Since any `enum` implements `Comparable`, it is not reasonable to say that the order may be relied on "unexpectedly".

What I mean is that not every enum client uses `compareTo()` or `values()[0]` and `values()[values().length - 1]` to get the opposite extremes of a certain enum class. Not every client uses `EnumSet.range`. However, given enough time, they likely, eventually will.

Reply:
I believe I addressed your concerns in 2def7bf.